src/CaptureClient/CaptureSaveToFileProcessor.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "CaptureClient/CaptureEventProcessor.h"
+#include "CaptureFile/CaptureFile.h"
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "OrbitBase/MakeUniqueForOverwrite.h"
+#include "capture_data.pb.h"
+
+using orbit_capture_file::CaptureFile;
+using orbit_capture_file::CaptureFileOutputStream;
+using orbit_client_protos::UserDefinedCaptureInfo;
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_capture_client {
+
+namespace {
+
+class CaptureSaveToFileProcessor : public CaptureEventProcessor {
+ public:
+  explicit CaptureSaveToFileProcessor(std::filesystem::path file_path,
+                                      absl::flat_hash_set<uint64_t> frame_track_function_ids,
+                                      std::function<void(const ErrorMessage&)> error_handler)
+      : frame_track_function_ids_(std::move(frame_track_function_ids)),
+        file_path_{std::move(file_path)},
+        error_handler_{std::move(error_handler)} {}
+  ~CaptureSaveToFileProcessor() override = default;
+
+  ErrorMessageOr<void> Initialize();
+  void ProcessEvent(const ClientCaptureEvent& event) override;
+
+ private:
+  ErrorMessageOr<void> WriteUserData();
+
+  absl::flat_hash_set<uint64_t> frame_track_function_ids_;
+  std::filesystem::path file_path_;
+  std::function<void(const ErrorMessage&)> error_handler_;
+  std::unique_ptr<CaptureFileOutputStream> output_stream_;
+};
+
+ErrorMessageOr<void> CaptureSaveToFileProcessor::Initialize() {
+  auto stream_or_error = CaptureFileOutputStream::Create(file_path_);
+  if (stream_or_error.has_error()) {
+    return ErrorMessage{absl::StrFormat("Failed to initialize CaptureSaveToFileProcessor: %s",
+                                        stream_or_error.error().message())};
+  }
+
+  output_stream_ = std::move(stream_or_error.value());
+
+  return outcome::success();
+}
+
+void CaptureSaveToFileProcessor::ProcessEvent(const ClientCaptureEvent& event) {
+  CHECK(output_stream_ != nullptr);
+  if (!output_stream_->IsOpen()) return;
```

Comment:
Could this be reported to the `error_handler_`? For example in case an event is received after the `CaptureFinished` event?

Reply:
The another case here is after I/O error happened. In this case reporting this to error_handler_ will result in spam. Note that error_handler_ is not resposible for cleaning up the file in case I/O happens it is done by the CaptureFileOutputStream. After first error ProcessEvent should ignore all other errors.

Reply:
Yeah, I guess also having multiple events after the `CaptureFinished` event would result in spam. But in both cases you could report the error only if it's the first time.

Reply:
ok.. I introduced additional state to catch "an event after CaptureFinished" error.